### ISSUE TITLE
Intermittent trouble with dist.libuv.org

### ISSUE THREAD

**Issue by johnsonjh** [libuv/libuv#4081]
I've been having some (intermittent) failures with the libuv website:

```
https://dist.libuv.org/dist/v1.45.0/libuv-v1.45.0.tar.gz:
2023-07-03 07:00:57 ERROR 500: Internal Server Error.
```

Just mentioning in case it's not already known.

**Comment by johnsonjh** [comment]
Now https://dist.libuv.org is completely down.

**Comment by bnoordhuis** [comment]
@saghul I believe you host libuv.org?

**Comment by saghul** [comment]
👋  Nope, it was some machine the Node build team let us use...

**Comment by johnsonjh** [comment]
It seems fine today; I will close and re-open if it recurs.

**Comment by Ambyjkl** [comment]
I'm also facing this, it has been occurring for several months now, but right now it's happening very frequently to me

**Comment by richardlau** [comment]
> 👋 Nope, it was some machine the Node build team let us use...

To confirm, while the Node.js Build WG does not manage the dist.libuv.org DNS entry, it is pointing to the same primary webserver that is used by nodejs.org, which has been rather unstable recently.

**Comment by saghul** [comment]
Should it point somewhere else?

**Comment by richardlau** [comment]
> Should it point somewhere else?

I'm fine with it pointing at the Node.js www server. Just pointing out it's kind of in the same boat as the nodejs.org server (or worse since there's no fall over server).

**Comment by saghul** [comment]
Oh. So is the Nodejs www server having some trouble?

I wonder if we could move dist somewhere else... GH pages maybe, since they are small assets? Thoughts @libuv/collaborators ?

**Comment by richardlau** [comment]
> Oh. So is the Nodejs www server having some trouble?

Unfortunately yes. It has got significantly worse this week 😞.

**Comment by bnoordhuis** [comment]
nodejs.org has been having trouble for at least two months now with no end in sight so I guess we should move somewhere else.

Apropos GH pages, from [here](https://docs.github.com/en/pages/getting-started-with-github-pages/about-github-pages#usage-limits):

1. sites may be no larger than 1 GB
2. sites have a soft bandwidth limit of 100 GB per month

I suspect we fit those criteria but I'm not entirely sure.

**Comment by richardlau** [comment]
I think that only dist.libuv.org is on the Node.js server. The libuv website (libuv.org) is somewhere else.
Refs: https://github.com/nodejs/build/blob/main/ansible/www-standalone/resources/config/libuv.org?plain=1

**Comment by richardlau** [comment]
Also FWIW the entirety of dist.libuv.org on the Node.js web server is 134 mb (yes, megabytes).

**Comment by bnoordhuis** [comment]
That's great. Some googling tells me GH Pages supports TLS for custom domains through Let's Encrypt so I think we should be good to go.

The only question left is... who manages the domain name? @saghul?

**Comment by vtjnash** [comment]
Do we have a mirror or backup of the content anywhere? If we could download the content, someone could probably set up a script to upload them as attachments to the release objects on github.

**Comment by saghul** [comment]
> That's great. Some googling tells me GH Pages supports TLS for custom domains through Let's Encrypt so I think we should be good to go.
> 
> 
> 
> The only question left is... who manages the domain name? @saghul?

Yep I have the DNS, happy to adjust once we agree on the way forward!

**Comment by saghul** [comment]
> Do we have a mirror or backup of the content anywhere? If we could download the content, someone could probably set up a script to upload them as attachments to the release objects on github.

That would be nice too.

Maybe we could have it in both places? Or maybe just GH releases going forward?

**Comment by richardlau** [comment]
> Do we have a mirror or backup of the content anywhere? If we could download the content, someone could probably set up a script to upload them as attachments to the release objects on github.

The Build WG can get the data for dist.libuv.org off the existing nodejs.org www server. The data is also backed up on our backup server (to avoid confusion backup here is a proper backup server and separate from the fail over server for nodejs.org (which dist.libuv.org is not copied to)).

**Comment by bnoordhuis** [comment]
I think we're all in agreement, just needs someone to step up? I have time next week but probably not this weekend.

> Maybe we could have it in both places? Or maybe just GH releases going forward?

Mild preference for keeping things the way they are, to avoid breaking (possibly hypothetical) download scripts.

**Comment by saghul** [comment]
Good point on download scripts, let's have it the way it is, but in a new home.

**Comment by gabrieljbaker** [comment]
We aren't able to build our application because of this issue... it's been causing quite a bit of stress. Looking forward to the fix.

**Comment by dsinni** [comment]
Our production deployments are failing due to this outage. Please, help!

**Comment by Ambyjkl** [comment]
you can try swapping out dist.libuv.org links with ones from here https://github.com/libuv/libuv/releases, that has worked for me

**Comment by johnsonjh** [comment]
I should mention that what we do in some (but not all) of our CI/CD jobs is building libuv from the source distribution that we download and verify from the website.  Other jobs get libuv from a Linux distribution, or from GitHub with git, or from GitHub by downloading a zip.

We do this to verify our documentation - if you can't actually download libuv from dist.libuv.org and build it, then we don't want to be telling users they should be doing that in documentation.

So, it's not a showstopper (because it eventually works), but we have to re-try the pipelines several times due to the failures, so it slows things down.

Of course, we can cache libuv locally, but that's not something that I want to make a habit of, because it just hides the problem.  Because of our process is how I noticed it was having trouble again.

**Comment by gabrieljbaker** [comment]
Yeah ^. We use the rust crate and would rather not download and build ourselves if we can avoid it.

**Comment by bnoordhuis** [comment]
I'm open to contracting if you're in a rush to get this fixed, otherwise it'll be when I have time this week.

**Comment by gengjiawen** [comment]
> nodejs.org has been having trouble for at least two months now with no end in sight so I guess we should move somewhere else.
> 
> Apropos GH pages, from [here](https://docs.github.com/en/pages/getting-started-with-github-pages/about-github-pages#usage-limits):
> 
> 1. sites may be no larger than 1 GB
> 2. sites have a soft bandwidth limit of 100 GB per month
> 
> I suspect we fit those criteria but I'm not entirely sure.

If it's only binary， we can also put in npm.

**Comment by saghul** [comment]
It's the source tarball. I don't think npm is a good fit.

**Comment by bnoordhuis** [comment]
@richardlau is there anywhere I can download the data? I can try spidering but dist.nodejs.org has been super flaky for me lately.

(Well, not exactly "lately"... https://github.com/nodejs/nodejs.org/issues/5302)

**Comment by saghul** [comment]
Had some time this morning and I tinkered with a script to generate a directory listing (since GH pages doesn't support it), but yeah can't get the tarballs...

**Comment by richardlau** [comment]
While the web server is flaky (or actually broken at the moment) you should be able to ssh into the server (`ssh libuv@dist.libuv.org`). The downloads are under `/home/libuv/dist`. If it makes it easier, I've just tar'ed up the entire `/home/libuv` into `/tmp/libuv.tar.gz` which you should be able to grab via scp 
```console
scp libuv@dist.libuv.org:/tmp/libuv.tar.gz .
```

@bnoordhuis 's key is already authorized for the `libuv` user (key was unlabelled -- I've now labelled it) and there are keys for @vtjnash and @santigimeno. There are six unknown keys -- they didn't seem to match any of https://github.com/saghul.keys so I've added those to `/home/libuv/.ssh/authorized_keys`.

**Comment by richardlau** [comment]
> 
> ```console
> scp libuv@dist.libuv.org:/tmp/libuv.tar.gz .
> ```
> 

hmm. While that worked for me a moment ago I having trouble doing this from another machine 😞. If you're having trouble I do have the tarball -- I can possibly upload it someone else if you can suggest somewhere suitable.

**Comment by vtjnash** [comment]
That looks good to me. I wonder if there is a better way to handle mtime though, since git checkout will reset them if someone else tries to clone this, and the directories are all reset to today's date already as well.

**Comment by saghul** [comment]
Alright, that was a bit easier than I thought. I added a metadata cache so now the dates show correctly even for directories.

**Comment by johnsonjh** [comment]
@saghul That's looks like a nice solution you worked out there

**Comment by saghul** [comment]
If nobody objects I can move that to the libuv org and edit the DNS record to point it to dist.libuv.org and we should be good to go.

@libuv/collaborators WDYT?

**Comment by bnoordhuis** [comment]
SGTM! I'd set the DNS record's TTL low for now, in case we need to switch back in a hurry.

**Comment by saghul** [comment]
Makes sense, will do!

**Comment by saghul** [comment]
Fixed! Please also see: https://github.com/libuv/libuv/issues/4088

The repo is here now: https://github.com/libuv/dist.libuv.org

The release utility will need an update to  push the new tarballs onto the repo. Ping @santigimeno ?

**Comment by santigimeno** [comment]
> Fixed! Please also see: #4088

Thanks. Great work!

> The release utility will need an update to push the new tarballs onto the repo. Ping @santigimeno ?

Will do for next release.

**Comment by saghul** [comment]
I'm going to close this, looks like we are good!
